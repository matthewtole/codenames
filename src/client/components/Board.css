.BoardWrapper {
  flex: 1;
}

.Board {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  padding: 1px;
}

.BoardRow {
  display: flex;
  flex-direction: row;
  height: 20%;
}
