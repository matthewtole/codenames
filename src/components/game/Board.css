.BoardWrapper {
  flex: 1 1;
}

.Board {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  padding: 5px;
}

.BoardRow {
  display: flex;
  flex-direction: row;
  height: 20%;
}

@media only screen and (max-width: 1024px) {
  .Board {
    padding: 3px;
  }
}
