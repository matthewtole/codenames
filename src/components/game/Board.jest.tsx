import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { BoardProps, BoardMode, Board } from './Board';
import { OriginalCards1 } from '../../fixtures/cards';

describe('Board', () => {
  const defaultProps: BoardProps = {
    cards: OriginalCards1,
    width: 5,
    height: 5,
    revealedCards: [],
    mode: BoardMode.CONTROLLER,
    isGameOver: false,
    onRevealCard: jest.fn(),
    onHighlightCard: jest.fn(),
  };

  it('renders correctly', () => {
    const tree = renderer.create(<Board {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
