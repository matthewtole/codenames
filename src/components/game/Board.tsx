import * as React from 'react';
import './Board.css';
import { Card as CardComponent } from './Card';
import {
  Card,
  Coordinate,
  Role,
  CoordinateValue,
} from '../../redux/game/types';

export enum BoardMode {
  VIEWER = 'VIEWER',
  CONTROLLER = 'CONTROLLER',
}

export interface BoardProps {
  width: number;
  height: number;
  cards: Card[];
  revealedCards: number[];
  mode: BoardMode;
  highlighted?: Coordinate;
  highlightedRow?: number;
  isGameOver: boolean;
  onRevealCard: (card: Coordinate) => void;
  onHighlightCard: (card?: Coordinate) => void;
}

export class Board extends React.PureComponent<BoardProps, {}> {
  constructor(props: BoardProps) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <div className="BoardWrapper" onClick={() => this.highlightCard()}>
        <div className="Board">
          {this.renderRowHighlight()}
          {Array(this.props.height)
            .fill(0)
            .map((_: number, row: CoordinateValue) => this.renderRow(row))}
        </div>
      </div>
    );
  }

  private renderRowHighlight() {
    if (this.props.highlightedRow === undefined) {
      return null;
    }
    const row = this.props.highlightedRow;
    return (
      <div className="Board__row-highlight-wrapper">
        <div
          className="Board__row-highlight"
          style={{ top: 0, height: `${row * 20}%` }}
        />
        <div
          className="Board__row-highlight"
          style={{
            top: `${(row + 1) * 20}%`,
            height: `${(5 - row + 1) * 20}%`,
          }}
        />
      </div>
    );
  }

  private renderRow = (row: CoordinateValue) => {
    return (
      <div className="BoardRow" key={`row_${row}`}>
        {Array(this.props.width)
          .fill(0)
          .map((__: number, col: CoordinateValue) => this.renderCard(row, col))}
      </div>
    );
  }

  private renderCard = (row: CoordinateValue, col: CoordinateValue) => {
    const index = this.coordinateToIndex({ row, col });
    return (
      <CardComponent
        key={`card_${row}_${col}`}
        word={this.getWord(index)}
        role={this.getRole(index)}
        revealed={this.isRevealed(index)}
        highlighted={this.isHighlighted({ row, col })}
        onHighlight={() => {
          this.highlightCard({ row, col });
        }}
        onReveal={() => {
          this.revealCard({ row, col });
        }}
        boardMode={this.props.mode}
        isGameOver={this.props.isGameOver}
      />
    );
  }

  private getWord = (index: number): string => {
    return this.props.cards[index].word;
  }

  private getRole = (index: number): Role => {
    return this.props.cards[index].role;
  }

  private isRevealed = (index: number): boolean => {
    return this.props.revealedCards.indexOf(index) >= 0;
  }

  private isHighlighted = (card: Coordinate): boolean => {
    if (this.props.highlighted === undefined) {
      return false;
    }
    return (
      this.props.highlighted.row === card.row &&
      this.props.highlighted.col === card.col
    );
  }

  private highlightCard = (card?: Coordinate) => {
    this.props.onHighlightCard(card);
  }

  private revealCard = (card: Coordinate) => {
    this.props.onRevealCard(card);
  }

  private coordinateToIndex(coordinate: Coordinate): number {
    return coordinate.row * this.props.width + coordinate.col;
  }
}
