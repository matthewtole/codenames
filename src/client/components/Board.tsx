import * as React from 'react';
import './Board.css';
import { Card } from './Card';
import { CardData, Coordinate, Role } from '../../shared/game';

export enum BoardMode {
  Viewer = 'viewer',
  Controller = 'controller',
}

export interface BoardProps {
  width: number;
  height: number;
  cards: CardData[];
  mode: BoardMode;
  highlighted?: Coordinate;
  onRevealCard: (card: Coordinate) => void;
  onHighlightCard: (card?: Coordinate) => void;
}

export interface BoardState {
  highlighted?: Coordinate;
}

export class Board extends React.Component<BoardProps, BoardState> {
  constructor(props: BoardProps) {
    super(props);
    this.state = {};
  }

  render() {
    const rows = [];
    for (let r = 0; r < this.props.height; r += 1) {
      const row = [];
      for (let c = 0; c < this.props.width; c += 1) {
        const card = { row: r, col: c };
        row.push((
          <Card
            key={`card_${r}_${c}`}
            word={this.getWord(card)}
            role={this.getRole(card)}
            revealed={this.isRevealed(card)}
            highlighted={this.isHighlighted(card)}
            onHighlight={() => { this.highlightCard(card); }}
            onReveal={() => { this.revealCard(card); }}
          />
        ));
      }
      rows.push(<div className="BoardRow" key={`row_${r}`}>{row}</div>);
    }
    const classes = ['Board'];
    switch (this.props.mode) {
      case BoardMode.Controller: classes.push('Board--controller'); break;
      case BoardMode.Viewer: classes.push('Board--viewer'); break;
      default: // do nothing!
    }
    return (
      <div className="BoardWrapper" onClick={() => this.highlightCard()}>
        <div className={classes.join(' ')}>{rows}</div>
      </div>
    );
  }

  getWord = (card: Coordinate): string => {
    return this.props.cards[this.coordinateToIndex(card)].word;
  }

  getRole = (card: Coordinate): Role => {
    return this.props.cards[this.coordinateToIndex(card)].role;
  }

  isRevealed = (card: Coordinate): boolean => {
    return this.props.cards[this.coordinateToIndex(card)].isRevealed;
  }

  isHighlighted = (card: Coordinate): boolean => {
    if (this.props.highlighted === undefined) { return false; }
    return this.props.highlighted.row === card.row && this.props.highlighted.col === card.col;
  }

  highlightCard = (card?: Coordinate) => {
    this.props.onHighlightCard(card);
  }

  revealCard = (card: Coordinate) => {
    this.props.onRevealCard(card);
  }

  private coordinateToIndex(coordinate: Coordinate): number {
    return coordinate.row * this.props.width + coordinate.col;
  }
}
