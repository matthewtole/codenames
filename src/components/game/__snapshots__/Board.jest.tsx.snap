// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board renders correctly 1`] = `
<div
  className="BoardWrapper"
  onClick={[Function]}
>
  <div
    className="Board"
  >
    <div
      className="BoardRow"
    >
      <div
        className="Card Card--controller"
        data-role="BYSTANDER"
        onClick={[Function]}
      >
        <div>
          King
          <i
            className="fa fa-child"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BLUE_SPY"
        onClick={[Function]}
      >
        <div>
          Bug
          <i
            className="fa fa-tint"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BLUE_SPY"
        onClick={[Function]}
      >
        <div>
          Loch Ness
          <i
            className="fa fa-tint"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BLUE_SPY"
        onClick={[Function]}
      >
        <div>
          Scale
          <i
            className="fa fa-tint"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="RED_SPY"
        onClick={[Function]}
      >
        <div>
          Figure
          <i
            className="fa fa-fire"
          />
        </div>
      </div>
    </div>
    <div
      className="BoardRow"
    >
      <div
        className="Card Card--controller"
        data-role="BLUE_SPY"
        onClick={[Function]}
      >
        <div>
          March
          <i
            className="fa fa-tint"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BYSTANDER"
        onClick={[Function]}
      >
        <div>
          Hood
          <i
            className="fa fa-child"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BLUE_SPY"
        onClick={[Function]}
      >
        <div>
          Trip
          <i
            className="fa fa-tint"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BYSTANDER"
        onClick={[Function]}
      >
        <div>
          Pan
          <i
            className="fa fa-child"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="RED_SPY"
        onClick={[Function]}
      >
        <div>
          Wind
          <i
            className="fa fa-fire"
          />
        </div>
      </div>
    </div>
    <div
      className="BoardRow"
    >
      <div
        className="Card Card--controller"
        data-role="RED_SPY"
        onClick={[Function]}
      >
        <div>
          Port
          <i
            className="fa fa-fire"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="RED_SPY"
        onClick={[Function]}
      >
        <div>
          Chick
          <i
            className="fa fa-fire"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BLUE_SPY"
        onClick={[Function]}
      >
        <div>
          Copper
          <i
            className="fa fa-tint"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BLUE_SPY"
        onClick={[Function]}
      >
        <div>
          Tube
          <i
            className="fa fa-tint"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BYSTANDER"
        onClick={[Function]}
      >
        <div>
          Chest
          <i
            className="fa fa-child"
          />
        </div>
      </div>
    </div>
    <div
      className="BoardRow"
    >
      <div
        className="Card Card--controller"
        data-role="BLUE_SPY"
        onClick={[Function]}
      >
        <div>
          Robin
          <i
            className="fa fa-tint"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="RED_SPY"
        onClick={[Function]}
      >
        <div>
          Band
          <i
            className="fa fa-fire"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="RED_SPY"
        onClick={[Function]}
      >
        <div>
          Fire
          <i
            className="fa fa-fire"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BYSTANDER"
        onClick={[Function]}
      >
        <div>
          Chair
          <i
            className="fa fa-child"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="ASSASSIN"
        onClick={[Function]}
      >
        <div>
          Cat
          <i
            className="fa fa-bullseye"
          />
        </div>
      </div>
    </div>
    <div
      className="BoardRow"
    >
      <div
        className="Card Card--controller"
        data-role="BLUE_SPY"
        onClick={[Function]}
      >
        <div>
          Undertaker
          <i
            className="fa fa-tint"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="RED_SPY"
        onClick={[Function]}
      >
        <div>
          Cover
          <i
            className="fa fa-fire"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BYSTANDER"
        onClick={[Function]}
      >
        <div>
          Suit
          <i
            className="fa fa-child"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="RED_SPY"
        onClick={[Function]}
      >
        <div>
          Roulette
          <i
            className="fa fa-fire"
          />
        </div>
      </div>
      <div
        className="Card Card--controller"
        data-role="BYSTANDER"
        onClick={[Function]}
      >
        <div>
          Chocolate
          <i
            className="fa fa-child"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
